packages/clerk-js/src/core/resources/Session.ts

```diff
@@ -369,27 +369,43 @@ export class Session extends BaseResource implements SessionResource {
 
     const tokenId = this.#getCacheId(template, organizationId);
 
-    const cachedEntry = skipCache ? undefined : SessionTokenCache.get({ tokenId }, leewayInSeconds);
+    const cacheResult = skipCache ? undefined : SessionTokenCache.get({ tokenId }, leewayInSeconds);
 
     // Dispatch tokenUpdate only for __session tokens with the session's active organization ID, and not JWT templates
     const shouldDispatchTokenUpdate = !template && organizationId === this.lastActiveOrganizationId;
 
-    if (cachedEntry) {
+    if (cacheResult) {
+      // SWR: If token is expiring soon, trigger background refresh (fire-and-forget)
+      if (cacheResult.needsRefresh) {
+        debugLogger.debug('Token expiring soon, triggering background refresh', { tokenId }, 'session');
+        void this.#refreshTokenInBackground(template, organizationId, tokenId, shouldDispatchTokenUpdate);
+      }
+
       debugLogger.debug(
-        'Using cached token (no fetch needed)',
+        'Using cached token',
         {
+          needsRefresh: cacheResult.needsRefresh,
           tokenId,
         },
         'session',
       );
-      const cachedToken = await cachedEntry.tokenResolver;
+      const cachedToken = await cacheResult.entry.tokenResolver;
```

Comment:
At first I thought this would accidentally await the _new_ background promise, since it was being written to the cache synchronously. Because `tokenCache` creates a new value on `set`, this is not the case as `cacheResult` will contain the old value.

This is a bit implicit though, one way to make it explicit is to something like grabbing a reference before:

```tsx
    if (cacheResult) {
      const tokenResolver = cacheResult.entry.tokenResolver;

      // ...

      const cachedToken = await tokenResolver;
```

Very much a NIT. I think there might be a bigger problem with reusing the `tokenResolver` for the background refresh though, see separate comment.